url29ff2e53f3_nicolas3470_Settlers_of_Catan_AI_tgz-pJ8-robotGameJ8/src/soc/util/SOCGameList.java

```diff
@@ -0,0 +1,551 @@
+ *<P>
+ * The client-side addGame methods allow game names to have a prefix which marks them
+ * as unjoinable by the client ({@link SOCGames#MARKER_THIS_GAME_UNJOINABLE}).
+ * If the game name has this prefix, its {@link GameInfo#canJoin} flag is set to false,
+ * as queried by {@link #isUnjoinableGame(String)}.  The prefix is stripped within addGame,
+ * and not stored as part of the game name in this list.
+ * Besides addGame, never supply this prefix to a SOCGameList method taking a game name;
+ * supply the game name without the prefix.
+ *
+ * @author Robert S. Thomas
+ */
+public class SOCGameList
+{
+    /** key = String, value = {@link GameInfo}; includes mutexes to synchronize game state access,
+     *  game options, and other per-game info
+     */
+    protected Hashtable gameInfo;
+
+    /** map of game names to {@link SOCGame} objects */
+    protected Hashtable gameData;
+
+    /** used with gamelist's monitor */
+    protected boolean inUse;
+
+    /**
+     * constructor
+     */
+    public SOCGameList()
+    {
+        gameInfo = new Hashtable();
+        gameData = new Hashtable();
+        inUse = false;
+    }
+
+    /**
+     * take the monitor for this game list; if we must wait, sleep up to 1000 ms between attempts.
+     */
+    public synchronized void takeMonitor()
+    {
+        // D.ebugPrintln("SOCGameList : TAKE MONITOR");
+
+        while (inUse)
+        {
+            try
+            {
+                wait(1000);
+            }
+            catch (InterruptedException e)
+            {
+                System.out.println("EXCEPTION IN takeMonitor() -- " + e);
+            }
+        }
+
+        inUse = true;
+    }
+
+    /**
+     * release the monitor for this game list
+     */
+    public synchronized void releaseMonitor()
+    {
+        // D.ebugPrintln("SOCGameList : RELEASE MONITOR");
+        inUse = false;
+        this.notify();
+    }
+
+    /**
+     * take the monitor for this game
+     *
+     * @param game  the name of the game
+     * @return false if the game has no mutex, or game not found in the list
+     */
+    public boolean takeMonitorForGame(String game)
+    {
+        // D.ebugPrintln("SOCGameList : TAKE MONITOR FOR " + game);
+
+        GameInfo info = (GameInfo) gameInfo.get(game);
+        if (info == null)
+        {
+            return false;
+        }
+        MutexFlag mutex = info.mutex;
+
+        if (mutex == null)
+        {
+            return false;
+        }
+
+        boolean done = false;
+
+        while (!done)
+        {
+            if (mutex == null)
+            {
+                return false;
+            }
+
+            synchronized (mutex)
+            {
+                if (mutex.getState() == true)
+                {
+                    try
+                    {
+                        mutex.wait(1000);
+                    }
+                    catch (InterruptedException e)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)